src/osd/OSD.cc

```diff
@@ -7277,6 +7277,17 @@ void OSD::handle_osd_map(MOSDMap *m)
     skip_maps = true;
   }
 
+  // wait for pgs to catch up
+  {
+    epoch_t num = last - first + 1;
+    if (cct->_conf->osd_map_cache_size > num &&
+	cct->_conf->osd_map_cache_size < last) {
+      epoch_t min = last - cct->_conf->osd_map_cache_size + 1;
+      dout(10) << __func__ << " waiting for pgs to consume " << min << dendl;
+      service.wait_min_pg_epoch(min);
```

Comment:
I don't think we can stall progress; the PG code never blocks on anything other than backend IO.

Generally speaking, we have two choices: (1) ingest maps without bound or throttling and let the PGs get arbitrarily behind, or (2) keep PGs caught up to the cache (but potentially have a queue of newer map updates, up to whatever the incoming message throttle size is).  I think (1) is better for performance (we ensure cache will work).  The behavior might be slightly more wonky for a laggy OSD (a tell command might be slow?) but I think the simplicity is worth it.

Reply:
Did you mean you think (2) is better for performance? Otherwise that doesn't make sense.

And yes, I think (2) is the way to go here. But I'm worried because even if it's not an actual deadlock, it means a slow PG can prevent active PGs from doing stuff like, well, peering. Which seems like a serious issue!

Reply:
Yeah I meant 2, sorry :)

It means that a slowly peering PG can prevent injest of new maps... which could prevent peering (and various other stuff, yeah).  This doesn't concern me in the general sense (the work that advmap does is very lightweight) *but* if something is stuck somewhere holding a PG lock (e.g., hung read, deadlock, bug) then everything else will eventually grind to a halt too because the map consumption is blocked.  Until we suicide due to the failed heartbeat on the stuck PG thread.

In that case, though, it seems like the right mitigation is to perhaps trigger a suicide earlier if it is holding up map consumption? Seems like a secondary concern, though ("we could have detected our internal deadlock sooner").  And easy to fix with a similar (perhaps shorter?) heartbeat on the main dispatch thread.